honeybadger-java/src/main/java/io/honeybadger/reporter/HoneybadgerReporter.java

```diff
@@ -0,0 +1,427 @@
+
+            final String trimmed = tag.trim();
+
+            if (!trimmed.isEmpty()) {
+                tagHashSet.add(trimmed);
+            }
+        }
+
+        return Collections.unmodifiableSet(tagHashSet);
+    }
+
+    @SuppressWarnings("LiteralClassName")
+    protected boolean supportsHttpServletRequest() {
+        try {
+            Class.forName("javax.servlet.http.HttpServletRequest");
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    @SuppressWarnings("LiteralClassName")
+    protected boolean supportsPlayHttpRequest() {
+        try {
+            Class.forName("play.mvc.Http", false, this.getClass().getClassLoader());
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    protected NoticeReportResult submitError(final Throwable error,
+                                             final io.honeybadger.reporter.dto.Request request,
+                                             final String message,
+                                             final Set<String> tags) {
+        final String errorClassName = error.getClass().getName();
+        if (errorClassName != null &&
+                getConfig().getExcludedClasses().contains(errorClassName)) {
+            return null;
+        }
+
+        final Notice notice = new Notice(getConfig());
+
+        if (request != null) {
+            final String reportedMessage;
+            if (message != null && !message.isEmpty()) {
+                reportedMessage = message;
+            } else {
+                reportedMessage = parseMessage(error);
+            }
+
+            NoticeDetails noticeDetails = new NoticeDetails(
+                    getConfig(), error, tags, reportedMessage);
+            notice.setRequest(request).setError(noticeDetails);
+        } else {
+            NoticeDetails noticeDetails = new NoticeDetails(getConfig(), error, tags);
+            notice.setError(noticeDetails);
+        }
+
+        String json;
+        try {
+            json = OBJECT_MAPPER.writeValueAsString(notice);
+        } catch (JsonProcessingException e) {
+            logger.error("JSON Serialization of the Notice Failed.", e);
+            logger.error("Original Error", error);
+            return null;
+        }
+
+        for (int retries = 0; retries < RETRIES; retries++) {
+            try {
+                HttpResponse response = sendToHoneybadger(json)
+                        .returnResponse();
+                int responseCode = response.getStatusLine().getStatusCode();
+
+                if (responseCode != HttpStatus.SC_CREATED) {
+                    logger.error("Honeybadger did not respond with the " +
+                                    "correct code. Response was [{}]. Retries={}",
+                            responseCode, retries);
+                } else {
+                    UUID id = parseErrorId(response);
```

Comment:
This could return `null` we should handle this such that a user is left with an `NullPointerException` without an understanding of why it happened.

Reply:
I'm not quite sure I want to throw an exception for that case; it seems like that would create more confusion, and would change the interface, requiring a bit of rethinking of how consuming classes work. If I'm understanding right, it would only happen with malformed JSON responses; other error responses result in retries or failover. The null value should be sufficient for log parsing.